File: src/ladym/operations/system2.py
"""System2 background consolidation cycle (SPEC §2.8)."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ..schema import Layer


@dataclass
class System2Report:
    consolidate: Any = None
    proceduralize: Any = None
    l5: Any = None
    l6: Any = None
    decay: Any = None
    skipped_llm_steps: bool = False


def _count_recent_episodes(engine, workspace: str) -> int:
    return sum(
        1
        for _ in engine.store.iter_memories(
            workspace=workspace, layer=Layer.EPISODIC.value
        )
    )


def run_system2_cycle(engine, *, workspace: str | None = None) -> System2Report:
    ws = workspace or engine.config.workspace
    report = System2Report()
    report.consolidate = engine.consolidate(workspace=ws)
    report.proceduralize = engine.proceduralize(workspace=ws)
    enough = (
        _count_recent_episodes(engine, ws) >= engine.config.system2.min_episodes_to_run
    )
    if enough:
        if hasattr(engine, "extract_mental_models"):
            report.l5 = engine.extract_mental_models(workspace=ws)
        if hasattr(engine, "predict_forward_intents"):
            report.l6 = engine.predict_forward_intents(workspace=ws)
    else:
        report.skipped_llm_steps = True
    report.decay = engine.decay(workspace=ws)
    return report

File: operations/system2.go
package operations

// System2Report is the outcome of one background consolidation cycle.
type System2Report struct {
	Consolidate     any
	Proceduralize   any
	L5              any
	L6              any
	Decay           any
	SkippedLLMSteps bool
}

// System2Runner is the interface the engine satisfies to run a System 2 cycle
// (kept as an interface so operations does not import engine).
type System2Runner interface {
	Consolidate(workspace string, since float64) (*ConsolidationReport, error)
	Proceduralize(workspace string, minClusterSize int) (*ProceduralizeReport, error)
	ExtractMentalModels(workspace string) (*L5ExtractionReport, error)
	PredictForwardIntents(workspace string) (*L6PredictionReport, error)
	Decay(workspace string, dryRun bool, maxAgeS, activationFloor float64) (*DecayReport, error)
	CountRecentEpisodes(workspace string) (int, error)
	MinEpisodesToRun() int
}

// RunSystem2Cycle runs one System 2 cycle through the runner.
func RunSystem2Cycle(runner System2Runner, workspace string) (*System2Report, error) {
	report := &System2Report{}
	cons, err := runner.Consolidate(workspace, 0)
	if err != nil {
		return nil, err
	}
	report.Consolidate = cons
	proc, err := runner.Proceduralize(workspace, 0)
	if err != nil {
		return nil, err
	}
	report.Proceduralize = proc
	n, err := runner.CountRecentEpisodes(workspace)
	if err != nil {
		return nil, err
	}
	if n >= runner.MinEpisodesToRun() {
		l5, err := runner.ExtractMentalModels(workspace)
		if err != nil {
			return nil, err
		}
		report.L5 = l5
		l6, err := runner.PredictForwardIntents(workspace)
		if err != nil {
			return nil, err
		}
		report.L6 = l6
	} else {
		report.SkippedLLMSteps = true
	}
	decay, err := runner.Decay(workspace, true, 0, 0)
	if err != nil {
		return nil, err
	}
	report.Decay = decay
	return report, nil
}
